lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,331 @@
+  checkClientRectsInsideParents = true,
+}) {
+  const {
+    overflowX,
+    overflowY,
+    display,
+    opacity,
+    visibility,
+  } = getComputedStyle(node);
+
+  if (allClientRectsEmpty(clientRects)) {
+    if (
+      (overflowX === 'hidden' && overflowY === 'hidden') ||
+      node.children.length === 0
+    ) {
+      return false;
+    }
+  }
+
+  if (
+    display === 'none' ||
+    visibility === 'hidden' ||
+    visibility === 'collapse' ||
+    (opacity && parseFloat(opacity) < 0.1)
+  ) {
+    return false;
+  }
+
+  if (display === 'block' || display === 'inline-block') {
+    if (node.clientWidth === 0 && overflowX === 'hidden') {
+      return false;
+    }
+    if (node.clientHeight === 0 && overflowY === 'hidden') {
+      return false;
+    }
+  }
+
+  if (checkClientRectsInsideParents) {
+    if (!isWithinAncestorsVisibleScrollArea(node, clientRects)) {
+      // Treating overflowing content in scroll containers as invisible could mean that
+      // most of a given page is deemed invisible. But:
+      // - tap targets audit doesn't consider different containers/layers
+      // - having most content in an explicit scroll container is rare
+      // - treating them as hidden only generates false passes, which is better than false failures
+      return false;
+    }
+  }
+
+  const parent = node.parentElement;
+  if (
+    parent &&
+    parent.tagName !== 'HTML' &&
+    !isVisible({node: parent, checkClientRectsInsideParents: false})
+  ) {
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * @param {Element} node
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {boolean}
+ */
+function isWithinAncestorsVisibleScrollArea(node, clientRects) {
+  const parent = node.parentElement;
+  if (!parent) {
+    return true;
+  }
+  if (getComputedStyle(parent).overflowY !== 'visible') {
+    for (let i = 0; i < clientRects.length; i++) {
+      const clientRect = clientRects[i];
+      if (!rectContains(parent.getBoundingClientRect(), clientRect)) {
+        return false;
+      }
+    }
+  }
+  if (parent.parentElement && parent.parentElement.tagName !== 'HTML') {
+    return isWithinAncestorsVisibleScrollArea(
+      parent.parentElement,
+      clientRects
+    );
+  }
+  return true;
+}
+
+/**
+ * @param {string} str
+ * @param {number} maxLength
+ * @returns {string}
+ */
+function truncate(str, maxLength) {
+  if (str.length <= maxLength) {
+    return str;
+  }
+  return str.slice(0, maxLength - 1) + '…';
+}
+
+/**
+ * @param {Element} node
+ * @param {boolean} includeChildren
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function getClientRects(node, includeChildren = true) {
+  const rawClientRects = /** @type {LH.Artifacts.ClientRect[]} */ Array.from(
+    node.getClientRects()
+  );
+
+  let clientRects = Array.from(rawClientRects).map(clientRect => {
+    const {width, height, left, top, right, bottom} = clientRect;
+    return {width, height, left, top, right, bottom};
```

Comment:
Yeah, will add a comment. I don't really understand why it's needed though, since calling JSON.stringify on a DOMRect is fine. But when we use `Runtime.evaluate` the clientRects come back empty (`{}`).